packages/eslint-plugin-circuit-ui/no-invalid-custom-properties/index.spec.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright 2023, SumUp Ltd.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// We disable the rule in this file because we explicitly test invalid cases
+/* eslint-disable @sumup/circuit-ui/no-invalid-custom-properties */
+
+import { ESLintUtils } from '@typescript-eslint/utils';
+
+import { noInvalidCustomProperties } from '.';
+
+const ruleTester = new ESLintUtils.RuleTester({
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    ecmaFeatures: {
+      jsx: true,
+    },
+  },
+});
+
+ruleTester.run('no-invalid-custom-properties', noInvalidCustomProperties, {
+  valid: [
+    {
+      name: 'custom properties in a JS object',
+      code: `
+        const COLOR_MAP = {
+          default: "var(--cui-fg-normal)",
+          active: "var(--cui-fg-normal-pressed)",
+        }
+      `,
+    },
+    {
+      name: 'custom properties in a tagged template literal',
+      code: `
+        const styles = css\`
+          color: var(--cui-fg-normal);
+          background-color: var(--cui-fg-normal-pressed);
+        \`;
+      `,
+    },
+    {
+      name: 'custom properties in inline styles',
+      code: `
+        function Component() {
+          return (
+            <p
+              style="color:var(--cui-fg-normal);background-color:var(--cui-fg-normal-pressed);"
+            >
+              Success
+            </p>
+          );
+        }
+      `,
+    },
+  ],
+  invalid: [
```

Comment:
Do you think it's worth adding a test case for interpolated variables (e.g. `var(--cui-fg-${variant})`)?

Reply:
Maybe, yes! But some notes about this:

- these will only be caught if they start with `--cui-` including the dash after "cui". I think this is fine because it would be weird for someone to have a map of colors starting with a `-`, but if we want to cover for it, I can also tweak the regex to match `--cui*` instead of `--cui-*`. What do you think?

    ```js
    const styles = css`
      /* caught */
      color: var(--cui-fg-${variant});
      color: var(--cui-${colorMap[variant]});
      /* not caught */
      color: var(--cui${colorMap[variant]});
    `;
    ```
- The ESLint error for these won't be very nice, because it will say something like `"--cui-fg-" is not a valid Circuit UI custom property`. It would be pretty difficult to make this nicer I think (would need a more complex regex or another rule). Are we okay with the error not being great as long as it flags occurrences properly?

Reply:
If we make this a separate rule, we'd still need to exclude these cases from the current role 🤔

Reply:
☝️ 

I was wrong, only the first example (the one you shared, too) would be caught. `var(--cui-${interpolation})` would still pass because the regex checks for `--cui-` followed by `\w-` (so the `$` doesn't match).

I'd still start by only covering that pattern and would look into tweaking the regex to catch other cases if we notice people using them (it'll be pretty easy to check using GitHub CodeSearch). Test case added in 3a2c6335b130920d696de54ee3a552af927d64b9

Reply:
Oops, didn't see your replies before posting my follow-up. But I think we're on the same page, and I agree about the separate rule. Let's look at this in a follow-up PR and release this 🙂